File: src/main/java/system/SettingsManager.java
package system;

import system.storage.StorageManager;

import java.io.*;
import java.util.Properties;
import java.util.logging.Logger;

/**
 * Used to save, retrieve and manage user settings.
 */
public class SettingsManager {
    public static final String SETTINGS_FILENAME = "settings.properties";

    // Settings properties
    public static final String ENABLE_DOKEY_SEARCH = "ENABLE_DOKEY_SEARCH";

    // Default settings
    public static final String DEFAULT_ENABLE_DOKEY_SEARCH = String.valueOf(true);

    private Properties properties = new Properties();
    private StorageManager storageManager;

    // Create the logger
    private final static Logger LOG = Logger.getGlobal();

    public SettingsManager(StorageManager storageManager) {
        this.storageManager = storageManager;

        if (!getSettingsFile().isFile()) {  // NO settings file available.
            // Default settings
            properties.setProperty(ENABLE_DOKEY_SEARCH, DEFAULT_ENABLE_DOKEY_SEARCH);

            // Save the default settings
            savePreferences();
        }else{  // File exist, load settings from it
            // Load the properties
            try {
                FileInputStream fis = new FileInputStream(getSettingsFile());
                properties.load(fis);
                fis.close();
            } catch (IOException e) {
                LOG.severe("Cannot load project settings");
            }
        }
    }

    /**
     * Save properties to file.
     * @return true if succeeded, false otherwise.
     */
    private boolean savePreferences() {
        try {
            FileOutputStream fos = new FileOutputStream(getSettingsFile());
            properties.store(fos, "Dokey Settings");
            fos.close();
            return true;
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return false;
    }

    /**
     * @return the file where settings are saved to.
     */
    private File getSettingsFile() {
        return new File(storageManager.getStorageDir(), SETTINGS_FILENAME);
    }

    // Specific properties

    /**
     * @return true if dokey search is enabled, false otherwise.
     */
    public boolean isDokeySearchEnabled() {
        return Boolean.valueOf(properties.getProperty(ENABLE_DOKEY_SEARCH, DEFAULT_ENABLE_DOKEY_SEARCH));
    }

    /**
     * Set the dokey search enabled property, save the settings and send the broadcast.
     * @param isEnabled
     * @return true if succeeded, false otherwise.
     */
    public boolean setDokeySearchEnabled(boolean isEnabled) {
        properties.setProperty(ENABLE_DOKEY_SEARCH, String.valueOf(isEnabled));
        // Send the broadcast
        BroadcastManager.getInstance().sendBroadcast(BroadcastManager.ENABLE_DOKEY_SEARCH_PROPERTY_CHANGED, isEnabled);
        return savePreferences();
    }
}

File: src/main/java/system/SettingsManager.kt
package system

import system.startup.StartupManager
import system.storage.StorageManager

import java.io.*
import java.util.Properties
import java.util.logging.Logger

const val SETTINGS_FILENAME = "settings.properties"

// Settings properties
const val ENABLE_DOKEY_SEARCH = "ENABLE_DOKEY_SEARCH"
const val SHOW_DELETED_COMMANDS = "SHOW_DELETED_COMMANDS"

// Default settings
const val DEFAULT_ENABLE_DOKEY_SEARCH = true.toString()
const val DEFAULT_SHOW_DELETED_COMMANDS = false.toString()

/**
 * Used to save, retrieve and manage user settings.
 */
class SettingsManager(storageManager: StorageManager) {
    // Create the logger
    private val LOG = Logger.getGlobal()

    private val properties = Properties()
    private val settingsFile = File(storageManager.storageDir, SETTINGS_FILENAME)

    // Specific properties

    /**
     * true if dokey search is enabled, false otherwise.
     */
    var dokeySearchEnabled : Boolean
        get() = properties.getProperty(ENABLE_DOKEY_SEARCH, DEFAULT_ENABLE_DOKEY_SEARCH).toBoolean()
        set(isEnabled) {
            properties.setProperty(ENABLE_DOKEY_SEARCH, isEnabled.toString())
            // Send the broadcast
            BroadcastManager.getInstance().sendBroadcast(BroadcastManager.ENABLE_DOKEY_SEARCH_PROPERTY_CHANGED, isEnabled)
            savePreferences()
        }

    var showDeletedCommands : Boolean
        get() = properties.getProperty(SHOW_DELETED_COMMANDS, DEFAULT_SHOW_DELETED_COMMANDS).toBoolean()
        set(isEnabled) {
            properties.setProperty(SHOW_DELETED_COMMANDS, isEnabled.toString())
            savePreferences()
        }

    init {

        if (!settingsFile.isFile) {  // NO settings file available.
            // Default settings
            properties.setProperty(ENABLE_DOKEY_SEARCH, DEFAULT_ENABLE_DOKEY_SEARCH)
            properties.setProperty(SHOW_DELETED_COMMANDS, DEFAULT_SHOW_DELETED_COMMANDS)

            // Save the default settings
            savePreferences()
        } else {  // File exist, load settings from it
            // Load the properties
            try {
                val fis = FileInputStream(settingsFile)
                properties.load(fis)
                fis.close()
            } catch (e: IOException) {
                LOG.severe("Cannot load project settings")
            }

        }
    }

    /**
     * Save properties to file.
     * @return true if succeeded, false otherwise.
     */
    private fun savePreferences(): Boolean {
        try {
            val fos = FileOutputStream(settingsFile)
            properties.store(fos, "Dokey Settings")
            fos.close()
            return true
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return false
    }
}
